webhook/csv.go

```diff
@@ -0,0 +1,141 @@
+	OperatorNamespace string
+
+	ctx context.Context
+	log logr.Logger
+}
+
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=clusterserviceversions,verbs=get;patch
+
+func (r *ClusterServiceVersionDefaulter) Handle(ctx context.Context, req admission.Request) admission.Response {
+
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+	r.log.Info("request received for csv review")
+
+	instance := &opv1a1.ClusterServiceVersion{}
+	if err := r.Decoder.Decode(req, instance); err != nil {
+		r.log.Error(err, "failed decoding admission review as csv")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed decoding admission review as csv: %v", err))
+	}
+
+	if ok := r.isOdfManagedCsv(instance); !ok {
+		r.log.Info("ignoring requested csv as it is not relevant")
+		return admission.Allowed("csv is not relevant to the odf")
+	}
+
+	replicas, err := r.getOldCsvReplicas(instance)
+	if err != nil {
+		r.log.Error(err, "failed getting replicas from the csv")
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed getting replicas from the csv: %v", err))
+	}
+
+	r.mutateCsv(instance, &replicas)
+
+	marshaledInstance, err := json.Marshal(instance)
+	if err != nil {
+		r.log.Error(err, "failed marshaling csv")
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed marshaling csv: %v", err))
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledInstance)
+}
+
+func (r *ClusterServiceVersionDefaulter) getOldCsvReplicas(instance *opv1a1.ClusterServiceVersion) (int32, error) {
+
+	if instance.Spec.Replaces == "" {
+		r.log.Info("csv.Spec.Replaces is not populated")
+		return 0, nil
+	}
+
+	oldCsv := &opv1a1.ClusterServiceVersion{}
+	key := client.ObjectKey{Name: instance.Spec.Replaces, Namespace: instance.Namespace}
+
+	if err := r.Client.Get(r.ctx, key, oldCsv); err != nil {
+		if errors.IsNotFound(err) {
+			// new install where an older csv does not exists
+			return 0, nil
+		}
+		r.log.Error(err, "failed getting older csv")
+		return 0, err
+	}
+
+	for _, deployment := range oldCsv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		if *deployment.Spec.Replicas > 0 {
+			// upgrade case where an older csv is found with replica 1
+			return 1, nil
+		}
+	}
+
+	// upgrade case where an older csv is found with replica 0
+	return 0, nil
+}
+
+func (r *ClusterServiceVersionDefaulter) mutateCsv(instance *opv1a1.ClusterServiceVersion, replicas *int32) {
+
+	r.log.Info("mutating requested csv")
+	for i := range instance.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		instance.Spec.InstallStrategy.StrategySpec.DeploymentSpecs[i].Spec.Replicas = replicas
+	}
+}
+
+func (r *ClusterServiceVersionDefaulter) isOdfManagedCsv(instance *opv1a1.ClusterServiceVersion) bool {
+
+	if instance.Namespace != r.OperatorNamespace {
+		return false
+	}
+
+	for _, pkgName := range controllers.PkgNames {
+		if strings.HasPrefix(instance.ObjectMeta.Name, pkgName) {
```

Comment:
ObjectMeta and TypeMeta is embedded into the struct and could directly use `.Name`

Reply:
Done